pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+				},
+					pod,
+				)
+				if err != nil {
+					if !kerrors.IsNotFound(err) {
+						return nil, err
+					}
+				} else {
+					if err := r.Delete(ctx, pod); err != nil {
+						return nil, err
+					}
+				}
+				log.Debugw("Migrating kubeone cluster container runtime", "from", currentContainerRuntime, "to", wantedContainerRuntime)
+				if pod, err = r.migrateCluster(ctx, log, externalCluster); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+	return pod, nil
+}
+
+func getWantedVersion(currentManifest []byte) (string, error) {
 	cluster := &kubeonev1beta2.KubeOneCluster{}
 	if err := yaml.UnmarshalStrict(currentManifest, cluster); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to decode manifest secret data: %w", err)
+		return "", fmt.Errorf("failed to decode manifest secret data: %w", err)
 	}
 
-	wantedVersion := cluster.Versions.Kubernetes
-	// check if a upgrade pod already exists
+	return cluster.Versions.Kubernetes, nil
+}
+
+func (r *reconciler) checkPodStatusIfExists(ctx context.Context,
+	log *zap.SugaredLogger,
+	externalCluster *kubermaticv1.ExternalCluster,
+	action string) error {
 	pod := &corev1.Pod{}
-	err = r.Get(ctx, ctrlruntimeclient.ObjectKey{
-		Name:      KubeOneUpgradePod,
+	var podName string
+	if action == UpgradeControlPlaneAction {
+		podName = KubeOneUpgradePod
+	} else if action == MigrateControlPlaneAction {
+		podName = KubeOneMigratePod
+	}
+
+	err := r.Get(ctx, ctrlruntimeclient.ObjectKey{
+		Name:      podName,
 		Namespace: kubernetesprovider.GetKubeOneNamespaceName(externalCluster.Name),
 	},
 		pod,
 	)
 	if err != nil {
-		if kerrors.IsNotFound(err) {
-			if externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status != kubermaticv1.StatusError {
-				if currentVersion != wantedVersion {
-					log.Debugw("Upgrading kubeone cluster", "from", currentVersion, "to", wantedVersion)
-					if err := r.upgradeCluster(ctx, log, externalCluster); err != nil {
-						return reconcile.Result{}, err
-					}
-				}
-			} else {
-				return reconcile.Result{}, nil
-			}
-		} else {
-			return reconcile.Result{}, err
+		if !kerrors.IsNotFound(err) {
+			return err
 		}
 	} else {
-		if pod.Status.Phase == corev1.PodPending || pod.Status.Phase == corev1.PodRunning {
-			return reconcile.Result{RequeueAfter: time.Second * 10}, nil
-		} else {
-			err = r.checkPodStatus(ctx, log, pod, externalCluster)
-			if err != nil {
-				return reconcile.Result{}, err
-			}
+		err = r.checkPodStatus(ctx, log, pod, externalCluster, action)
+		if err != nil {
+			return err
 		}
 	}
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, externalCluster *kubermaticv1.ExternalCluster) (reconcile.Result, error) {
+	if err := r.initiateImportAction(ctx, log, externalCluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterStatus := externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status
```

Comment:
This PR doesn't introduce it, but why do we keep status innformation in the spec? This line looks confusing to me... can you explain why we have status in the spec?

Reply:
For any operation to perform for kubeone cluster, we are verifying if the cluster status is `Running`.
this will help in 2 things:
- avoiding 2 parallel reconciling(upgrading/migrating) operations being performed.
- avoiding operation on a cluster with an error status.
So we will trigger a new operation only when the Cluster is in `Running` (!Reconciling/!Error) status

Reply:
That doesn't explain why status information is in the spec.

Reply:
Where should we place the status?

Reply:
In the status subresource: https://book-v1.book.kubebuilder.io/basics/status_subresource.html -- the same thing we use for Project status, Cluster status, User status and (soon) Seed status :)